python_modules/dagster/dagster/core/execution/with_resources.py

```diff
@@ -3,26 +3,88 @@
 from dagster import _check as check
 from dagster.utils import merge_dicts
 
+from ...config import Shape
 from ..definitions import ResourceDefinition
 from ..definitions.resource_requirement import ResourceAddable
+from ..errors import DagsterInvalidConfigError, DagsterInvalidInvocationError
 
 T = TypeVar("T", bound=ResourceAddable)
 
 
 def with_resources(
     definitions: Iterable[T],
     resource_defs: Mapping[str, ResourceDefinition],
-    config: Optional[Dict[str, Any]] = None,
+    resource_config_by_key: Optional[Mapping[str, Any]] = None,
 ) -> Sequence[T]:
+    """Function that provides dagster resources to resource-requiring dagster definitions.
```

Comment:
I think it's worth being explicit that this returns copies of the existing definitions, doesn't modify them. Similar to how we document `group_name` on `load_assets_from_package_module` etc: https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/core/asset_defs/load_assets_from_modules.py#L84